cpp/path_module/algorithm/path.cpp

```diff
@@ -1,301 +1,343 @@
-                                : RelationshipAllowed(rel_type, relationshipSets, any_outgoing, any_incoming, false);
-    if (!rel_allowed) {  // if relationship not allowed, go to next one
+
+    auto record = record_factory.NewRecord();
+    record.Insert(std::string(kResultCreate).c_str(), path);
+
+  } catch (const std::exception &e) {
+    record_factory.SetErrorMessage(e.what());
+    return;
+  }
+}
+
+void Path::PathExpand::ExpandPath(mgp::Path &path, const mgp::Relationship &relationship, int64_t path_size) {
+  path.Expand(relationship);
+  visited_.insert(relationship.Id().AsInt());
+  DFS(path, path_size + 1);
+  visited_.erase(relationship.Id().AsInt());
+  path.Pop();
+}
+
+void Path::PathExpand::Expand(mgp::Path &path, mgp::Relationships relationships, bool outgoing, int64_t path_size,
+                              std::set<std::pair<std::string_view, int64_t>> &seen) {
+  for (const auto relationship : relationships) {
+    auto type = std::string(relationship.Type());
+    auto wanted_direction = helper_.GetDirection(type);
+
+    if ((wanted_direction == RelDirection::kNone && !helper_.AnyDirected(outgoing)) ||
+        visited_.contains(relationship.Id().AsInt())) {
       continue;
     }
-    cpy.Expand(rel);  // expand the path with this relationships
-    std::unordered_set<mgp::Relationship> relationships_set_cpy = relationships_set;
-    relationships_set_cpy.insert(rel);  // insert the relationship into visited relationships
-
-    /*this part is for label filtering*/
-    LabelBools labelBools;
-    mgp::Labels labels = outgoing ? rel.To().Labels() : rel.From().Labels();
-    for (auto label : labels) {  // set booleans to their value for the label of the finish node
-      FilterLabel(label, labelSets, labelBools);
-    }
 
-    if (PathSizeOk(path_size, max_hops, min_hops) && ShouldExpand(labelBools, labelStatus)) {
-      auto record = record_factory.NewRecord();
-      record.Insert(std::string(std::string(kResultExpand).c_str()).c_str(), cpy);
-    }
-    if (path_size + 1 < max_hops && !labelBools.blacklisted && !labelBools.terminated &&
-        (labelBools.end_node || Whitelisted(labelBools.whitelisted, labelStatus.whitelist_empty))) {
-      PathDFS(cpy, relationships_set_cpy, record_factory, path_size + 1, min_hops, max_hops, labelSets, labelStatus,
-              relationshipSets, any_outgoing, any_incoming);
+    RelDirection curr_direction = outgoing ? RelDirection::kOutgoing : RelDirection::kIncoming;
+
+    if (wanted_direction == RelDirection::kAny || curr_direction == wanted_direction || helper_.AnyDirected(outgoing)) {
+      ExpandPath(path, relationship, path_size);
+    } else if (wanted_direction == RelDirection::kBoth) {
+      if (outgoing && seen.contains({type, relationship.To().Id().AsInt()})) {
+        ExpandPath(path, relationship, path_size);
+      } else {
+        seen.insert({type, relationship.From().Id().AsInt()});
+      }
```

Comment:
Can you explain logic a bit here too

Reply:
Same logic as the previous practically same piece of code.

Reply:
Does it make sense to isolate this part of code @imilinovic then? In function or something

Reply:
I think not really because this part can also have some other stuff in it, potentionally bfs and dfs could be isolated later for when ExpandConfig will be implemented.